velox/type/Conversions.h

```diff
@@ -28,159 +28,214 @@ DECLARE_bool(experimental_enable_legacy_cast);
 
 namespace facebook::velox::util {
 
-template <
-    TypeKind KIND,
-    typename = void,
-    bool TRUNCATE = false,
-    bool LEGACY_CAST = false>
+template <TypeKind KIND, typename = void>
 struct Converter {
-  // TODO: simplify template parameters and evaluate performance impact.
   template <typename T>
-  static typename TypeTraits<KIND>::NativeType cast(T) {
+  typename TypeTraits<KIND>::NativeType cast(T) {
     VELOX_UNSUPPORTED(
         "Conversion to {} is not supported", TypeTraits<KIND>::name);
   }
+
+  bool truncate = false;
+  bool legacyCast = false;
 };
 
-template <bool LEGACY_CAST>
-struct Converter<TypeKind::BOOLEAN, void, false, LEGACY_CAST> {
+template <>
+struct Converter<TypeKind::BOOLEAN> {
   using T = bool;
 
   template <typename From>
-  static T cast(const From& v) {
+  T cast(const From& v) {
+    if (truncate) {
+      VELOX_UNSUPPORTED("Conversion to BOOLEAN is not supported");
+    }
     return folly::to<T>(v);
   }
 
-  static T cast(folly::StringPiece v) {
+  T cast(folly::StringPiece v) {
     return folly::to<T>(v);
   }
 
-  static T cast(const StringView& v) {
+  T cast(const StringView& v) {
     return folly::to<T>(folly::StringPiece(v));
   }
 
-  static T cast(const std::string& v) {
+  T cast(const std::string& v) {
+    return folly::to<T>(v);
+  }
+
+  T cast(const bool& v) {
     return folly::to<T>(v);
   }
 
-  static T cast(const Timestamp&) {
+  T cast(const float& v) {
+    if (truncate) {
+      if (std::isnan(v)) {
+        return false;
+      }
+      return v != 0;
+    } else {
+      return folly::to<T>(v);
+    }
```

Comment:
Should it be 
```
if (std::isnan(v)) {
        VELOX_USER_FAIL("Cannot cast NaN to an integral value.");
      }
      return folly::to<T>(std::round(v));
```
to remain the same behavior?

Reply:
Hi @kagamiori  Wei, I double checked, and this logic branch (To Boolean, TRUNCATE=false) should come from line 51 instead of line 244, before the change. This is the confusing part and I try to refactor it out.

Previously the "To Boolean" case is divided into two structs in this file. For the TRUNCATE=false branch, it's line 46
```
struct Converter<TypeKind::BOOLEAN, void, false, LEGACY_CAST>
```
So we should get logic here from this struct.

For the TRUNCATE=true branch, it's line 72
```
template <TypeKind KIND, bool TRUNCATE, bool LEGACY_CAST>
struct Converter<
    KIND,
    std::enable_if_t<
        KIND == TypeKind::BOOLEAN || KIND == TypeKind::TINYINT ||
            KIND == TypeKind::SMALLINT || KIND == TypeKind::INTEGER ||
            KIND == TypeKind::BIGINT || KIND == TypeKind::HUGEINT,
        void>,
    TRUNCATE,
    LEGACY_CAST>
```

I also double checked a unit test for this branch, copying here
```
    // When TRUNCATE = false.
    testConversion<float, bool>(
        {
            0.1,
            0.0,
            -0.1,
            kInf,
            kNan,
        },
        {
            true,
            false,
            true,
            true,
            true,
        },
        /*truncate*/ false);
```